### DISCUSSION TITLE
Task - TypeScript - Library Management System

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/angular#147] [Tasks]
You have been hired to create a library management system for a small local library. The library wants to keep track of their books and authors, as well as provide a simple way to add new books and authors to their system.

Create two classes, `Author` and `Book`, with the following specifications:

`Author` class:

- Properties:
  `name`: string - The name of the author.

`birthYear`: number - The birth year of the author.

- Constructor:
  Accepts two arguments, `name` and `birthYear`, and initializes the respective properties.

`Book` class:

- Properties:
  `title`: string - The title of the book.

`author`: Author - The author of the book.

`publicationYear`: number - The publication year of the book.

- Constructor:
  Accepts three arguments, `title`, `author`, and `publicationYear`, and initializes the respective properties.

Create a few instances of the `Author` and `Book` classes to simulate adding new authors and books to the library's collection. Then, display the information about the books and their authors in a human-readable format.

Example:

```
const author1 = new Author("George Orwell", 1903);
const book1 = new Book("1984", author1, 1949);

console.log(`Title: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);

```

Output:

```
Title: 1984, Author: George Orwell, Publication Year: 1949

```

**Comment by bipin-coded** [comment]
### Code
```typescript
class Author {
  name: string;
  birthYear: number;

  constructor(name: string, birthYear: number) {
    this.name = name;
    this.birthYear = birthYear;
  }
}

class Book {
  title: string;
  author: Author;
  publicationYear: number;

  constructor(title: string, author: Author, publicationYear: number) {
    this.title = title;
    this.author = author;
    this.publicationYear = publicationYear;
  }
}

const author1 = new Author("George Orwell", 1903);
const book1 = new Book("1984", author1, 1949);

console.log(`Title: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);
```

### Output
![Screenshot from 2023-04-23 16-47-34](https://user-images.githubusercontent.com/126976134/233849957-6c37d0ab-426d-4914-b7d5-4ff9062a3098.png)

**Comment by sam2elect** [comment]
# CODE
```ts
class AuthorTS {
    name: string = '';
    birthYear: number = 0;
  
    constructor(name: string = '', birthYear: number = 0) {
      this.name = name;
      this.birthYear = birthYear;
    }
  }
  
  class BookTS {
    title: string = '';
    author: Author = new Author();
    publicationYear: number = 0;
  
    constructor(title: string = '', author: Author = new Author(), publicationYear: number = 0) {
      this.title = title;
      this.author = author;
      this.publicationYear = publicationYear;
    }
  }
  
  const author1 = new AuthorTS("George Orwell", 1903);
  const book1 = new BookTS("1984", author1, 1949);
  console.log(`Title: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);
  
  
```

# OUTPUT

![image](https://user-images.githubusercontent.com/85131631/233849883-e0cfb759-148f-46cb-a73a-61761359f5dc.png)

**Comment by kumarneeraj89** [comment]
# Code

```javascript
class Book2Ts{

    title: string;
    author: AuthorTs;
    publicationYear: number;

    constructor(title: string,author: AuthorTs, publicationYear: number){
        this.title  = title;
        this.author = author;
        this.publicationYear = publicationYear;
    }

    displayInfo(){
        return `Title: ${this.title}, Author: + ${this.author}.`
    }
};

class AuthorTs{

    name: string;
    birthYear: number;

    constructor(name: string,birthYear: number){
        this.name = name;
        this.birthYear = birthYear;
    }
}

let author1ts = new AuthorTs("George Orwell", 1903)
const book2ts = new Book2Ts("1984", author1ts, 1949)

console.log(`Title: ${book2ts.title}, Author: ${book2ts.author.name}, Publication Year: ${book2ts.publicationYear}`);


```

**Comment by SathyaGanes** [comment]
```ts
class Author {
  name: string;
  birthYear: number;

  constructor(name: string, birthYear: number) {
    this.name = name;
    this.birthYear = birthYear;
  }
}

class Book {
  title: string;
  author: Author;
  publicationYear: number;

  constructor(title: string, author: Author, publicationYear: number) {
    this.title = title;
    this.author = author;
    this.publicationYear = publicationYear;
  }
}

const author1 = new Author("George Orwell", 1903);
const book1 = new Book("1984", author1, 1949);

console.log(`Title: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);

```

OUTPUT:
<img width="1124" alt="Screenshot 2023-04-23 at 9 22 46 PM" src="https://user-images.githubusercontent.com/126976887/233850292-adb33c1f-16d2-4a30-9db4-8f92a0fccd6d.png">

**Comment by Prathmesh06** [comment]
# Code
```js
class AuthorTs {
    name : string;
    birthYear : number;

    constructor(name: string, birthYear: number) {
        this.name = name;
        this.birthYear = birthYear;
    }

}

const author1 = new AuthorTs("George Orwell", 1903);

class BookTs {
    title : string;
    author : AuthorTs;
    publicationYear : number;

    constructor(title : string, author : AuthorTs, publicationYear : number) {
        this.title = title;
        this.author = author;
        this.publicationYear = publicationYear;
    }
}

const book1 = new BookTs("1984", author1, 1949);
console.log(book1);
```

# output
![image](https://user-images.githubusercontent.com/58461727/233850267-d6073506-58a8-433b-83e0-f941bcf86ff8.png)
![image](https://user-images.githubusercontent.com/58461727/233850319-f0fe407b-8f34-477a-a016-09fa26e7e665.png)

**Comment by basu-cse** [comment]
class Author {
    public name: string;
    public birthYear: number;
  
    constructor(name: string, birthYear: number) {
      this.name = name;
      this.birthYear = birthYear;
    }
}
  
class Book {
    public title: string;
    public author: Author;
    public publicationYear: number;
  
    constructor(title: string, author: Author, publicationYear: number) {
      this.title = title;
      this.author = author;
      this.publicationYear = publicationYear;
    }
  
    public getBookInfo(): string {
      return `${this.title} by ${this.author.name}, published in ${this.publicationYear}`;
    }
}

const author1 = new Author("George Orwell", 1903);

console.log(author1);

const book1 = new Book("1984", author1, 1949);
console.log(book1.getBookInfo());

Output
==============

![library](https://user-images.githubusercontent.com/95092878/233850569-1e81443f-744f-454c-babe-7e488b38b009.png)

**Comment by RevathiSundaramoorthy** [comment]
## Code
```typescript
class AuthorTS {
    name: string;
    birthYear: number;
    constructor(name: string, birthYear:number) {
        this.name = name; 
        this.birthYear = birthYear;  
    }
}

class BookTS {
    title: string;
    author: AuthorTS;
    publicationYear: number;
    constructor(title: string,author: AuthorTS,publicationYear: number) {
        this.title = title;
        this.author = author;
        this.publicationYear = publicationYear;
    }
    getBookInfo(): string {
        return ("Title: " + this.title + ", Author: " + this.author.name + ", Publication Year: " + this.publicationYear);
    }
}

const bookts = new BookTS("The Alchemist", new AuthorTS("Paulo Coelho", 1947), 1988);
console.log(bookts.getBookInfo());
const bookts1 = new BookTS("1984", new AuthorTS("George Orwell", 1903), 1949);
console.log(bookts1.getBookInfo());
```
## Screenshot
![image](https://user-images.githubusercontent.com/126975777/233850392-9bb0a492-7b48-4a58-87d4-7adb35a1a3bd.png)

**Comment by AratiKumari123** [comment]
```typescript
class Author{
  name:string;
  birthYear:number;
    
  constructor(name:string,birthYear:number){
    this.name=name;
    this.birthYear=birthYear;
  }
}

const Author1=new Author("George Orwell", 1903);
console.log(`name: ${author1.title}, birthYear: ${Author1.birthYear}`);
    
    
    class Book{
    title:string;
    author:Author;
    publicationYear:number;
    
    constructor(title:string,author:Author,publicationYear:number){
        this.title=title;
        this.author=author;
        this.publicationYear=publicationYear;
    }

}

const book1 = new Book("1984", author1, 1949);
console.log(`Title: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);
```

**Comment by VishalShukla85** [comment]
class AuthorTS  {
    name:string;
    brithYear:number;
    constructor(name:string, brithYear:number){
        this.name = name;
        this.brithYear = brithYear;
    }

}
class BookTS {
    title: string;
    author: AuthorTS;
    publications: number
    constructor(title:string, author: AuthorTS, publications: number){
        this.title = title;
        this.author = author;
        this.publications = publications;
    }
}
const author1 = new AuthorTS("The Guide", 1958);
const book1 = new BookTS("1984", author1, 1949);

console.log(`Title: ${book1.title}, 
Author: ${book1.author.name}, Publication Year: ${book1.publications}`);

**Comment by pujita789** [comment]
# code
class AuthorTs{
name:string;
birthYear:number;

constructor(name:string,birthYear:number){
this.name=name;
this.birthYear=birthYear;
}

}
class BookTs{
title:string;
author:AuthorTs;
publicationYear:number;
constructor(title:string,author:AuthorTs,publicationYear:number){
this.title=title;
this.author=author;
this.title=publicationYear;

}
}

const author1 = new AuthorTs("George Orwell", 1903);
const book1 = new BookTs("1984", author1, 1949);

console.log(`Title: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);

**Comment by vk0027** [comment]
class Author {
  name: string;
  birthYear: number;

  constructor(name: string, birthYear: number) {
    this.name = name;
    this.birthYear = birthYear;
  }
}

class Book {
  title: string;
  author: Author;
  publicationYear: number;

  constructor(title: string, author: Author, publicationYear: number) {
    this.title = title;
    this.author = author;
    this.publicationYear = publicationYear;
  }
}

const author1 = new Author("George Orwell", 1903);
const book1 = new Book("1984", author1, 1949);

console.log(`Title: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);

**Comment by mownica-p** [comment]
# code
```TypeScript
class Author {
    name: string;
    birthYear: number;
  
    constructor(name: string, birthYear: number) {
      this.name = name;
      this.birthYear = birthYear;
    }
  }
  
  class Book {
    title: string;
    author: Author;
    publicationYear: number;
  
    constructor(title: string, author: Author, publicationYear: number) {
      this.title = title;
      this.author = author;
      this.publicationYear = publicationYear;
    }
  }
 
//  instances of authors
const author1 = new Author('J.K. Rowling', 1965);
const author2 = new Author('George R.R. Martin', 1948);

//  instances of books
const book1 = new Book('Harry Potter and the Philosopher\'s Stone', author1, 1997);
const book2 = new Book('A Game of Thrones', author2, 1996);
const book3 = new Book('Harry Potter and the Chamber of Secrets', author1, 1998);

// display book and author information
console.log(`Book: ${book1.title}, Author: ${book1.author.name}, Publication Year: ${book1.publicationYear}`);
console.log(`Book: ${book2.title}, Author: ${book2.author.name}, Publication Year: ${book2.publicationYear}`);
console.log(`Book: ${book3.title}, Author: ${book3.author.name}, Publication Year: ${book3.publicationYear}`);
```
# ScreenShot
![image](https://user-images.githubusercontent.com/126975660/234894283-08b926a2-2cc1-410e-acf2-f8bafb58d879.png)

**Comment by Ravi-710** [comment]
/* You have been hired to create a library management system for a small local library.
The library wants to keep track of their books and authors, as well as provide a simple way to add new books and authors to their system.
Create two classes, Author and Book, with the following specifications: */

class Author_Ts{
  Name :string;
  Birthyear:number;
  constructor(Name:string, Birthyear:number){
        this.Name=Name;
        this.Birthyear=Birthyear;
  }
}
 class Book_Ts{
  Title:string;
  Author:Author_Ts;
  PublicationYear:Number;
  constructor(Title:string,Author:Author_Ts,PublicationYear:Number){
        this.Title=Title;
        this.Author=Author;
        this.PublicationYear=PublicationYear;
  }
 }
const Author_Obj=new Author_Ts("George",1903);
const Book_obj=new Book_Ts("1984",Author_Obj, 1949);

console.log(`Title:${Book_obj.Title}, Author:${Book_obj.Author.Name}, Pulication_Year:${Book_obj.PublicationYear}`);

**Comment by Ruchiarth** [comment]
```ts
//Task - TypeScript - Library Management System · akash-coded/angular · Discussion #147 (github.com) 


// Create two classes, Author and Book, with the following specifications:

// Author class:

// Properties:
// name: string - The name of the author.
// birthYear: number - The birth year of the author.

// Constructor:
// Accepts two arguments, name and birthYear, and initializes the respective properties.
class Author_ts {

        name :string;
        birthYear : number;

        constructor(name:string,birthYear:number)
        {
            this.name= name;
            this.birthYear = birthYear;
        }

}

// Book class:

// Properties:
// title: string - The title of the book.
// author: Author - The author of the book.

// publicationYear: number - The publication year of the book.

// Constructor:
// Accepts three arguments, title, author, and publicationYear, and initializes the respective properties.
class Book_ts {
    
    title :string;
    author :Author_ts ;
    publicationYear:number;

    constructor(title:string,author:Author_ts , publicationYear:number)
    {
        this.title = title;
        this.author = author;
        this.publicationYear=publicationYear;

    }

}

// Create a few instances of the Author and Book classes to simulate adding 
// new authors and books to the library's collection.

// Then, display the
//  information about the books and their authors in a human-readable format.
const author1 = new Author_ts("ChetanBhgarat",1908);
const author2 = new Author_ts("AS",1981);

const book1published = new Book_ts("MEF",author1,1990);
const book2published  = new Book_ts("MESC",author2,1989);

console.log(
    `Title: ${book1published.title}, Author: ${book1published.author.name} written in Year ${book1published.author.birthYear} ,with Publication Year as ${book1published.publicationYear}`);

console.log(
    `Title: ${book2published.title}, Author: ${book2published.author.name} written in Year ${book2published.author.birthYear} ,with Publication Year as ${book2published.publicationYear}`);

```
#Screenshot
![image](https://user-images.githubusercontent.com/126975575/235930100-250a327b-b9f8-4d7f-9319-a37ba467a02c.png)
